lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -356,18 +354,24 @@ static function (Query\Parameter $parameter) use ($key): bool {
      */
     public function setParameters($parameters)
     {
-        if (is_array($parameters)) {
-            /** @psalm-var ArrayCollection<int, Parameter> $parameterCollection */
-            $parameterCollection = new ArrayCollection();
-
-            foreach ($parameters as $key => $value) {
-                $parameterCollection->add(new Parameter($key, $value));
-            }
+        if (! is_array($parameters)) {
+            Deprecation::trigger(
+                'doctrine/orm',
+                'https://github.com/doctrine/orm/pull/9816',
+                'Passing an ArrayCollection to %s is deprecated.',
+                __METHOD__
+            );
+            $this->parameters = $parameters;
 
-            $parameters = $parameterCollection;
+            return $this;
         }
 
-        $this->parameters = $parameters;
+        $this->parameters = new ArrayCollection();
+
+        foreach ($parameters as $key => $value) {
+            $parameter = $value instanceof Parameter ? $value : new Parameter($key, $value);
+            $this->parameters->add($parameter);
```

Comment:
Before this PR, passing an array of `Parameter` instances does nothing special, so technically, this is a BC break.

Reply:
So we basically accept arrays that look like this?

```php
[
    'foo' => 'some value',
    new Parameter('bar', 'another value'),
]
```

Is this what we want?

Maybe we should split this method into two, one that accepts associative arrays and one that accepts a list of `Parameter` instances?

Reply:
Mmmh, I take back my proposal. It would work here, but it would complicate all other methods like `toIterable()`…

Reply:
> Is this what we want?

Your example shows a side-effect of allowing this, but it's not really realistic, is it? Anyway, I don't think we should check and throw in case the array is a mix of a hash and a list.

Reply:
Take all the time you need :slightly_smiling_face: